### PR TITLE
[JIT] Improve test coverage for ErrorReport instances

### PR DESCRIPTION
**Pull request by jamesr66a**
Baseline is here: https://github.com/pytorch/pytorch/issues/8634

New coverage for `test/test_jit.py`:`(line number, # calls, code)`:

```
torch/csrc/jit/script/compiler.cpp:
  148|      1|        throw ErrorReport(loc) << "cannot re-assign '" << name << "' to a value of type " << value->kind()
  153|      1|        throw ErrorReport(loc) << "cannot re-assign '" << name << "' because it has type " << value->kind()
  157|      2|        throw ErrorReport(loc) << "variable '" << name << "' previously has type " << simple_parent->type()->name()
  192|      4|      throw ErrorReport(range) << "undefined value " << ident;
  475|      1|      throw ErrorReport(loc) << "argument 'chunks' must be a constant";
  527|      1|    throw ErrorReport(loc) << "unknown builtin op";
  529|     11|  throw ErrorReport(loc) << "arguments for call are not valid:\n"
  543|      1|    throw ErrorReport(range) << "expected a tensor value but found a tuple";
  594|      1|        throw ErrorReport(def.params().range()) << "methods must have a self argument";
  691|      1|          throw ErrorReport(stmt) << "return statements can appear only at the end "
  891|      1|      throw ErrorReport(range)
  904|      1|      throw ErrorReport(stmt)
  908|      1|      throw ErrorReport(stmt) << "Iteration variable unpacking is not supported";
  912|      1|      throw ErrorReport(targets[0]) << "Starred unpacking is currently not"
  970|      1|        throw ErrorReport(assignee)
  976|      1|      throw ErrorReport(r)
  981|      1|      throw ErrorReport(r) << "A Starred expression may only appear on the "
  993|      1|        throw ErrorReport(stmt)
 1020|      3|      throw ErrorReport(stmt)
 1026|      4|      throw ErrorReport(stmt)
 1038|      1|          throw ErrorReport(var) << "Cannot pack a tuple into a non-variable.";
 1147|      1|        throw ErrorReport(ident) << "print doesn't accept any keyword arguments";
 1227|      0|        throw ErrorReport(tree) << "Unexpected starred expansion. File a bug report.";
 1262|      0|        throw ErrorReport(tree) << "NYI: " << tree;
 1283|      0|        throw ErrorReport(input) << "Unrecognized type: " << type;
 1342|      0|      throw ErrorReport(loc)
 1471|      4|  throw ErrorReport(loc) << value->type()->name() << " cannot be used as a tuple";
 1476|      1|    throw ErrorReport(loc) << "expected " << expected << " " << what << " but found " << actual;
torch/csrc/jit/script/compiler.h:
   41|      1|    throw ErrorReport(loc) << kind() << " cannot be used as a value";
   46|      1|    throw ErrorReport(loc) << "attribute lookup is not defined on " << kind();
   52|      1|    throw ErrorReport(loc) << kind() << " cannot be used as a tuple";
   78|      1|    throw ErrorReport(loc) << "cannot call a " << kind();
torch/csrc/jit/script/init.cpp:
   63|      3|      throw ErrorReport(loc) << "calling a Python function with an incorrect number "
   68|      1|        throw ErrorReport(loc) << "type mismatch at argument " << i << ": expected "
   77|      1|      throw ErrorReport(loc) << "keyword arguments in Python calls aren't supported";
  110|      1|        throw ErrorReport(loc) << "Python functions can currently only return Tensors";
  141|      1|      throw ErrorReport(loc) << "object has no attribute " << name;
  218|      2|  throw ErrorReport(loc) << "unsupported attribute lookup on " << py::repr(self) << ".";
  284|      2|        throw ErrorReport(loc) << "attribute '" << field << "' of type '" << typeString(attr) << "' is not usable in a script method (did you forget to add it __constants__?)";
  287|      0|    throw ErrorReport(loc) << "module has no attribute '" << field << "'";
torch/csrc/jit/script/module.cpp:
   34|      1|    throw ErrorReport(loc) << " method '" << callee.name()
   44|      1|    throw ErrorReport(loc) << failure_messages.str();
torch/csrc/jit/script/tree_views.h:
  169|      0|      throw ErrorReport(tree) << "Maybe trees can have at most one subtree";
  211|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid Type";
  228|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid Stmt";
  266|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid Expr";
  333|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid ScalarType";
  451|      0|        throw ErrorReport(tree) << "is not a valid AssignKind";
  525|      0|          throw ErrorReport(tree) << "BinOp expected 2 subtrees, found " << tree->trees().size();
  528|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid BinOp";
  548|      0|          throw ErrorReport(tree) << "UnaryOp expected 1 subtree, found " << tree->trees().size();
  551|      0|        throw ErrorReport(tree) << kindToString(tree->kind()) << " is not a valid UnaryOp";
```

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -3117,6 +3117,213 @@ def fn(x, y):
 
         self.checkScript(fn, (torch.randn(3, 2, dtype=torch.float), torch.ones(3, 2, dtype=torch.float)))
 
+    def test_reassign_module_lhs(self):
+        with self.assertRaisesRegex(RuntimeError, 'cannot re-assign \'self\' because it has type value. Only reassignments to first-class values are allowed'):
```
Comment (zdevito) [inline test/test_jit.py]: This error message is confusing. We should fix that (in a new issue)

File: test/test_jit.py [hunk 2]
```diff
@@ -3117,6 +3117,213 @@ def fn(x, y):
...
+        with self.assertRaisesRegex(RuntimeError, 'argument \'chunks\' must be a constant'):
+            @torch.jit.script
+            def chunk_non_constant(x, y):
+                return x.chunk(y)
+
+    def test_unknown_builtin(self):
+        with self.assertRaisesRegex(RuntimeError, 'unknown builtin op'):
+            @torch.jit.script
+            def unknown_builtin(x):
+                return x.splork(3)
```
Comment (zdevito) [inline test/test_jit.py]: nice name!

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -908,10 +908,6 @@ struct to_ir {
       throw ErrorReport(stmt) << "Iteration variable unpacking is not supported";
     }
 
-    if (targets[0].kind() != TK_VAR) {
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: was this impossible to hit?

**Review by zdevito — APPROVED** [review]
Cool!

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (jamesr66a) [reply torch/csrc/jit/script/compiler.cpp]: Actually, on second thought I think we would be able to hit this from the string frontend (if the string frontend supported starred expressions). Let me add this back and make a new issue

**Comment by ezyang** [comment]
@Jorghi12 ROCM failure here looks like a flaky build. Are we OOMing?

**Comment by ezyang** [comment]
Are you sure you don't want to use expect tests here? Obviously you'll have to do some legwork to expunge things like the stack traces, but if you ever make changes to the error messages (which you almost certainly will want to do), having `--accept` will be a lot more pleasant.

If you really want the messages inline in the file, I can code up a new version of `--accept` that can handle this.

**Comment by jamesr66a** [comment]
@ezyang hmm interesting. For example one output of these failures could be something like:

```
======================================================================
ERROR: test_wrong_return_type (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 3373, in test_wrong_return_type
    @torch.jit.script
  File "/Users/jamesreed/onnx-fairseq/pytorch/torch/jit/__init__.py", line 371, in script
    graph = _script_graph(fn, _frames_up=_frames_up + 1)
  File "/Users/jamesreed/onnx-fairseq/pytorch/torch/jit/__init__.py", line 367, in _script_graph
    return _jit_script_compile(ast, rcb)
RuntimeError:
Python functions can currently only return Tensors:
@torch.jit.script
def wrong_return_type():
    return somefunc()
           ~~~~~~~~ <--- HERE


----------------------------------------------------------------------
Ran 1 test in 0.111s

FAILED (errors=1)
```

Maybe we can make a version of expect which can regex match? So maybe we could match this substring:

```
RuntimeError:
Python functions can currently only return Tensors:
@torch.jit.script
def wrong_return_type():
    return somefunc()
           ~~~~~~~~ <--- HERE
```

Which is probably more stable wrt not having absolute paths or anything. How does that sound?

**Comment by ezyang** [comment]
Yes, that would be great.

**Comment by jamesr66a** [comment]
Actually, I don't think that's going to work because we're going to have to manually specify the match string anyway, unless we can figure out how to automatically strip out the stack trace stuff. @ezyang what would `assertRaisesRegex` look like if it supported accept files/`--accept`?

**Comment by ezyang** [comment]
@jamesr66a A common way to do it is to write a normalizer: it takes some string and normalizes it into a new string which resembles the original, but is intended to be invariant across different environments. You could implement it as a regex, but instead of having a regex which *matches modulo system differences*, you have a regex which matches system differences and removes them.

Another way to do it is refactor the error reporting so that the stack is separate from the message, so you can just expect test the message.

**Pull request merged**